src/topics/fork.js

```diff
@@ -10,171 +10,116 @@ var plugins = require('../plugins');
 var meta = require('../meta');
 
 module.exports = function (Topics) {
-	Topics.createTopicFromPosts = function (uid, title, pids, fromTid, callback) {
+	Topics.createTopicFromPosts = async function (uid, title, pids, fromTid) {
 		if (title) {
 			title = title.trim();
 		}
 
 		if (title.length < meta.config.minimumTitleLength) {
-			return callback(new Error('[[error:title-too-short, ' + meta.config.minimumTitleLength + ']]'));
+			throw new Error('[[error:title-too-short, ' + meta.config.minimumTitleLength + ']]');
 		} else if (title.length > meta.config.maximumTitleLength) {
-			return callback(new Error('[[error:title-too-long, ' + meta.config.maximumTitleLength + ']]'));
+			throw new Error('[[error:title-too-long, ' + meta.config.maximumTitleLength + ']]');
 		}
 
 		if (!pids || !pids.length) {
-			return callback(new Error('[[error:invalid-pid]]'));
+			throw new Error('[[error:invalid-pid]]');
 		}
 
-		pids.sort(function (a, b) {
-			return a - b;
-		});
+		pids.sort((a, b) => a - b);
+
 		var mainPid = pids[0];
-		var cid;
-		var tid;
-		async.waterfall([
-			function (next) {
-				posts.getCidByPid(mainPid, next);
-			},
-			function (_cid, next) {
-				cid = _cid;
-				async.parallel({
-					postData: function (next) {
-						posts.getPostData(mainPid, next);
-					},
-					isAdminOrMod: function (next) {
-						privileges.categories.isAdminOrMod(cid, uid, next);
-					},
-				}, next);
-			},
-			function (results, next) {
-				if (!results.isAdminOrMod) {
-					return next(new Error('[[error:no-privileges]]'));
-				}
-				Topics.create({ uid: results.postData.uid, title: title, cid: cid }, next);
-			},
-			function (tid, next) {
-				Topics.updateTopicBookmarks(fromTid, pids, function (err) { next(err, tid); });
-			},
-			function (_tid, next) {
-				tid = _tid;
-				async.eachSeries(pids, function (pid, next) {
-					privileges.posts.canEdit(pid, uid, function (err, canEdit) {
-						if (err || !canEdit.flag) {
-							return next(err || new Error(canEdit.message));
-						}
-
-						Topics.movePostToTopic(uid, pid, tid, next);
-					});
-				}, next);
-			},
-			function (next) {
-				Topics.updateLastPostTime(tid, Date.now(), next);
-			},
-			function (next) {
-				plugins.fireHook('action:topic.fork', { tid: tid, fromTid: fromTid, uid: uid });
-				Topics.getTopicData(tid, next);
-			},
-		], callback);
+		var cid = await posts.async.getCidByPid(mainPid);
+
+		const [postData, isAdminOrMod] = await Promise.all([
+			posts.async.getPostData(mainPid),
+			privileges.async.categories.isAdminOrMod(cid, uid),
+		]);
+
+		if (!isAdminOrMod) {
+			throw new Error('[[error:no-privileges]]');
+		}
+		const tid = await Topics.create({ uid: postData.uid, title: title, cid: cid });
+		await Topics.updateTopicBookmarks(fromTid, pids);
+
+		await async.eachSeries(pids, async function (pid) {
```

Comment:
This is just a for...of loop in an async function.

Reply:
Does for of work in node 8?

Reply:
Yes.

Reply:
Our eslint rules are complaing about await inside loop :rage1: